IoT-Sample/Swift/IoTSampleSwift/ConnectionViewController.swift

```diff
@@ -19,207 +19,273 @@ import AWSMobileClient
+                // Connect to the AWS IoT data plane service w/ certificate
+                iotDataManager.connect( withClientId: uuid, cleanSession:true, certificateId:certificateId!, statusCallback: self.mqttEventCallback)
+            }
+        } else {
+            handleDisconnect()
+            DispatchQueue.main.async {
+                sender.setTitle("Connect \(self.IOT_CERT)", for:UIControl.State())
+                sender.isEnabled = true
+            }
+        }
+    }
 
-                    // Now create and store the certificate ID in NSUserDefaults
-                    let csrDictionary = [ "commonName":CertificateSigningRequestCommonName, "countryName":CertificateSigningRequestCountryName, "organizationName":CertificateSigningRequestOrganizationName, "organizationalUnitName":CertificateSigningRequestOrganizationalUnitName ]
-
-                    self.iotManager.createKeysAndCertificate(fromCsr: csrDictionary, callback: {  (response ) -> Void in
-                        if (response != nil)
-                        {
-                            defaults.set(response?.certificateId, forKey:"certificateId")
-                            defaults.set(response?.certificateArn, forKey:"certificateArn")
-                            certificateId = response?.certificateId
-                            print("response: [\(String(describing: response))]")
-
-                            let attachPrincipalPolicyRequest = AWSIoTAttachPrincipalPolicyRequest()
-                            attachPrincipalPolicyRequest?.policyName = PolicyName
-                            attachPrincipalPolicyRequest?.principal = response?.certificateArn
-                            
-                            // Attach the policy to the certificate
-                            self.iot.attachPrincipalPolicy(attachPrincipalPolicyRequest!).continueWith (block: { (task) -> AnyObject? in
-                                if let error = task.error {
-                                    print("failed: [\(error)]")
-                                }
-                                print("result: [\(String(describing: task.result))]")
-                                
-                                // Connect to the AWS IoT platform
-                                if (task.error == nil)
-                                {
-                                    DispatchQueue.main.asyncAfter(deadline: .now()+2, execute: {
-                                        self.logTextView.text = "Using certificate: \(certificateId!)"
-                                        self.iotDataManager.connect( withClientId: uuid, cleanSession:true, certificateId:certificateId!, statusCallback: mqttEventCallback)
+    func handleDisconnect() {
+        let tabBarViewController = tabBarController as! IoTSampleTabBarController
+        self.connectButton.isHidden = false
+        self.connectIoTDataWebSocket.isHidden = false
+        activityIndicatorView.startAnimating()
```

Comment:
Do we ever stop animating?

Reply:
Yes, the callback should get called either mqttEventCallback or mqttEventCallbackWebsocket